### PR TITLE
[Feature] Support DeepSeek MTP on NPU

### PR DESCRIPTION
**Pull request by iforgetmyname**
## Motivation

This pr primarily aims to support deepseek's mtp on ascend npus.

## Modifications

- Introduces NPU support for newest eagle framework
- Includes ascend specific ops for draft tree build/verify

## Accuracy Tests

pr

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Alcanderian** [comment]
Can we add an PR Test for NPU MTP？

**Comment by Yellowhappy** [comment]
Hi, which DeepSeek model is this, and is it running on one machine or two?

**Comment by iforgetmyname** [comment]
> Hi, which DeepSeek model is this, and is it running on one machine or two?

Hi, this supports both V3 and V3.2, and it could run on one machine if hbm capacity allows

**Comment by iforgetmyname** [comment]
> Can we add an PR Test for NPU MTP？

for sure, we have test_ascend_deepseek_mtp.py for pr-test now

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/srt/speculative/eagle_info.py [hunk 1]
```diff
@@ -133,16 +137,25 @@ def prepare_for_verify(self, batch: ScheduleBatch, page_size: int):
...
-        assign_req_to_token_pool[(bs,)](
-            batch.req_pool_indices,
-            batch.req_to_token_pool.req_to_token,
-            batch.seq_lens,
-            end_offset,
-            batch.out_cache_loc,
-            batch.req_to_token_pool.req_to_token.shape[1],
-            next_power_of_2(bs),
-        )
+        if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info.py]: _is_npu, and please check other codes

File: python/sglang/srt/speculative/eagle_info.py [hunk 2]
```diff
@@ -77,18 +77,22 @@ def get_spec_adjust_token_coefficient(self) -> Tuple[int, int]:
 
     @classmethod
     def create_idle_input(cls, topk: int, spec_steps: int, num_verify_tokens: int):
+        if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info.py [hunk 3]
```diff
@@ -276,20 +290,37 @@ def verify(
...
-                predicts=predict,  # mutable
-                accept_index=accept_index,  # mutable
-                accept_token_num=accept_length,  # mutable
-                candidates=candidates,
-                retrive_index=self.retrive_index,
-                retrive_next_token=self.retrive_next_token,
-                retrive_next_sibling=self.retrive_next_sibling,
-                target_predict=target_predict,
-            )
+            if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info.py [hunk 4]
```diff
@@ -468,15 +501,24 @@ def verify(
...
-                assign_req_to_token_pool[(bs,)](
-                    batch.req_pool_indices,
-                    batch.req_to_token_pool.req_to_token,
-                    batch.seq_lens,
-                    batch.seq_lens + accept_length + 1,
-                    batch.out_cache_loc,
-                    batch.req_to_token_pool.req_to_token.shape[1],
-                    next_power_of_2(bs),
-                )
+                if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info_v2.py [hunk 5]
```diff
@@ -112,15 +112,27 @@ def prepare_for_decode(self: EagleDraftInput, batch: ScheduleBatch):
...
-        assign_req_to_token_pool[(bs,)](
-            batch.req_pool_indices,
-            batch.req_to_token_pool.req_to_token,
-            self.allocate_lens,
-            new_allocate_lens,
-            out_cache_loc,
-            batch.req_to_token_pool.req_to_token.shape[1],
-            next_power_of_2(bs),
-        )
+        if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info_v2.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info.py [hunk 6]
```diff
@@ -468,15 +501,24 @@ def verify(
...
+                    torch.ops.npu.cache_loc_assign(
+                        batch.req_pool_indices,
+                        batch.req_to_token_pool.req_to_token,
+                        batch.seq_lens,
+                        batch.seq_lens + accept_length + 1,
+                        batch.out_cache_loc,
+                    )
+                else:
+                    assign_req_to_token_pool[(bs,)](
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info.py]: Can we abstract a unify interface for `assign_req_to_token_pool`?

And then replace the symbol for npu, example:
```
def assign_req_to_token_pool_triton(...)
    # call assign_req_to_token_pool[(bs,)] in this interface

assign_req_to_token_pool = assign_req_to_token_pool_triton
if _is_npu:
    assign_req_to_token_pool = assign_req_to_token_pool_npu
```

File: python/sglang/srt/speculative/eagle_info_v2.py [hunk 7]
```diff
@@ -199,21 +211,36 @@ def prepare_for_v2_verify(
         bs = len(batch.req_pool_indices)
         batch.input_ids = self.draft_token
         device = batch.input_ids.device
-        batch.out_cache_loc = torch.empty(
-            (bs * self.draft_token_num,),
-            dtype=torch.int64,
-            device=device,
-        )
+        if is_npu():
+            batch.out_cache_loc = torch.empty(
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info_v2.py]: same as assign_req_to_token_pool

File: python/sglang/srt/speculative/eagle_info_v2.py [hunk 8]
```diff
@@ -199,21 +211,36 @@ def prepare_for_v2_verify(
         bs = len(batch.req_pool_indices)
         batch.input_ids = self.draft_token
         device = batch.input_ids.device
-        batch.out_cache_loc = torch.empty(
-            (bs * self.draft_token_num,),
-            dtype=torch.int64,
-            device=device,
-        )
+        if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info_v2.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info_v2.py [hunk 9]
```diff
@@ -258,20 +285,36 @@ def sample(
...
-                predicts=predict,  # mutable
-                accept_index=accept_index,  # mutable
-                accept_token_num=accept_length,  # mutable
-                candidates=candidates,
-                retrive_index=self.retrive_index,
-                retrive_next_token=self.retrive_next_token,
-                retrive_next_sibling=self.retrive_next_sibling,
-                target_predict=target_predict,
-            )
+            if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info_v2.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info_v2.py [hunk 10]
```diff
@@ -258,20 +285,36 @@ def sample(
...
+                    target_predict,
+                    accept_index,
+                    accept_length,
+                    predict,
+                    self.draft_token_num,
+                    self.topk,
+                )
+
+            else:
+                verify_tree_greedy(
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_info_v2.py]: same as assign_req_to_token_pool

File: python/sglang/srt/speculative/eagle_utils.py [hunk 11]
```diff
@@ -114,20 +224,41 @@ def build_tree_kernel_efficient(
...
-        positions,
-        retrive_index,
-        retrive_next_token,
-        retrive_next_sibling,
-        topk,
-        spec_steps,
-        num_verify_tokens,
-        tree_mask_mode,
-    )
+    if is_npu():
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_utils.py]: _is_npu

File: python/sglang/srt/speculative/eagle_utils.py [hunk 12]
```diff
@@ -114,20 +224,41 @@ def build_tree_kernel_efficient(
...
+            tree_mask,
+            retrive_index,
+            retrive_next_token,
+            retrive_next_sibling,
+            topk,
+            num_verify_tokens,
+            tree_mask_mode,
+            bs,
+        )
+    else:
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_utils.py:251]: same as assign_req_to_token_pool

File: python/sglang/srt/speculative/eagle_worker.py [hunk 13]
```diff
@@ -940,7 +941,7 @@ def capture_for_decode(
         draft_input.hidden_states = logits_output.hidden_states
 
 
-@torch.compile(dynamic=True)
+@torch.compile(dynamic=True, disable=is_npu())
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/eagle_worker.py]: _is_npu

File: python/sglang/srt/speculative/spec_utils.py [hunk 14]
```diff
@@ -331,7 +333,7 @@ def get_target_cache_loc(
     )
 
 
-@torch.compile(dynamic=True)
+@torch.compile(dynamic=True, disable=is_npu())
```
Comment (Alcanderian) [inline python/sglang/srt/speculative/spec_utils.py]: _is_npu

File: python/sglang/srt/speculative/eagle_info.py (hunk 6 above)
Comment (iforgetmyname) [reply python/sglang/srt/speculative/eagle_info.py]: a unified interface requires both changing `sgl-kernel` and `sgl-kernel-npu`, so that currently i'm providing a grouped interface that dispatches to different hardwares underhood, let's see if this being a good short-term solution for now

**Review by hnyls2002 — APPROVED** [review]

**Comment by wangtiance** [comment]
Hello, is deepseek the only model supporting speculative decoding on NPU? Will qwen3 etc. be supported?
